### PR TITLE
Pv2 1544 fix on prog reporting funding line type

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Mapping/MapperConfigurationTest.cs [hunk 1]
```diff
@@ -117,7 +117,8 @@ public void Setup()
                         TotalNegotiatedPrice1 = 25.0m,
                         TotalNegotiatedPrice2 = 25.0m,
                         TotalNegotiatedPrice3 = 25.0m,
-                        TotalNegotiatedPrice4 = 25.0m
+                        TotalNegotiatedPrice4 = 25.0m,
+                        FundingLineType = "19+ Apprenticeship Levy Contract (procured)"
```
Comment (benagain) [inline src/SFA.DAS.Payments.DataLocks.Application.UnitTets/Mapping/MapperConfigurationTest.cs]: 🙃 If commas are included at the end of multiline lists, then future diffs won't highlight the line as changed when someone adds another item

File: src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Mapping/EarningEventMappingTest.cs [hunk 2]
```diff
@@ -256,6 +273,38 @@ public void PriceEpisodeMapsEarningsInfo(Type requiredPaymentEventType)
             requiredPaymentEvent.NumberOfInstalments.Should().Be((short)priceEpisode.NumberOfInstalments);
         }
 
+        [Test]
+        [TestCase(typeof(CalculatedRequiredIncentiveAmount))]
+        [TestCase(typeof(CalculatedRequiredCoInvestedAmount))]
+        [TestCase(typeof(CalculatedRequiredLevyAmount))]
+        public void MapPriceEpisodeToLearningAimFundingLineType(Type requiredPaymentEventType)
```
Comment (benagain) [inline src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Mapping/EarningEventMappingTest.cs]: 🔧 Is the parameter required?  It doesn't appear to be used in the test.

File: src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs [hunk 3]
```diff
@@ -234,5 +234,93 @@ public async Task TestHoldingBackCompletionPayment()
...
+                    new PriceEpisode
+                    {
+                        Identifier = "1",
+                        EffectiveTotalNegotiatedPriceStartDate = DateTime.UtcNow,
+                        PlannedEndDate = DateTime.UtcNow,
+                        ActualEndDate = DateTime.UtcNow,
+                        CompletionAmount = 100M,
+                        InstalmentAmount = 200M,
+                        NumberOfInstalments = 16,
+                        FundingLineType = "19+ Apprenticeship Non Levy Contract (procured)"
```
Comment (carlbm) [inline src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs]: Non-Levy has a hyphen

File: src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs [hunk 4]
```diff
@@ -234,5 +234,93 @@ public async Task TestHoldingBackCompletionPayment()
...
+                LearnAimReference = earningEvent.LearningAim.Reference,
+                TransactionType = (int) OnProgrammeEarningType.Learning
+            }};
+
+            paymentHistoryCacheMock.Setup(c => c.TryGet(It.Is<string>(key => key == CacheKeys.PaymentHistoryKey), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new ConditionalValue<PaymentHistoryEntity[]>(true, paymentHistoryEntities))
+                .Verifiable();
+            requiredPaymentsService.Setup(p => p.GetRequiredPayments(It.IsAny<Earning>(), It.IsAny<List<Payment>>()))
+                .Returns(requiredPayments)
+                .Verifiable();
```
Comment (carlbm) [inline src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs]: Why verify this when you don't care what's passed in?

File: src/SFA.DAS.Payments.RequiredPayments.Application/Mapping/RequiredPaymentsProfile.cs [hunk 5]
```diff
@@ -315,6 +315,17 @@ public RequiredPaymentsProfile()
                 .Ignore(x => x.AgreedOnDate)
                 ;
 
+            CreateMap<PriceEpisode, LearningAim>()
```
Comment (carlbm) [inline src/SFA.DAS.Payments.RequiredPayments.Application/Mapping/RequiredPaymentsProfile.cs]: I can't believe that this mapping is ever a good idea. I think that you may need to talk me through this reasoning

Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.RequiredPayments.Application/Mapping/RequiredPaymentsProfile.cs]: Already discussed with Tony, the alternative is a serious change, anyways happy to discuss tomorrow, perhaps can have a dev demo to explain the changes.

File: src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs (hunk 4 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs]: Is just to make sure the service is called

File: src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs (hunk 3 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs]: Done

File: src/SFA.DAS.Payments.DataLocks.Application/Mapping/DataLocksProfile.cs [hunk 6]
```diff
@@ -212,8 +212,10 @@ public DataLocksProfile()
                 .ForMember(dest => dest.CompletionAmount, opt => opt.MapFrom(source => source.CompletionAmount))
                 .ForMember(dest => dest.Completed, opt => opt.MapFrom(source => source.Completed))
                 .ForMember(dest => dest.EmployerContribution, opt => opt.MapFrom(source => source.EmployerContribution))
-                .ForMember(dest => dest.CompletionHoldBackExemptionCode, opt => opt.MapFrom(source => source.CompletionHoldBackExemptionCode));
-            
+                .ForMember(dest => dest.CompletionHoldBackExemptionCode, opt => opt.MapFrom(source => source.CompletionHoldBackExemptionCode))
+                .ForMember(dest => dest.FundingLineType, opt => opt.Ignore());
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Application/Mapping/DataLocksProfile.cs]: changed

File: src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs (hunk 4 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.RequiredPayments.Application.UnitTests/Application/Processors/PayableEarningEventProcessorTest.cs]: But the rest of the assertions wouldn't work if it wasn't called. Isn't this just noise?

File: src/SFA.DAS.Payments.RequiredPayments.Application/Mapping/RequiredPaymentsProfile.cs (hunk 5 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.RequiredPayments.Application/Mapping/RequiredPaymentsProfile.cs]: I'll accept this under duress. The mapping doesn't make sense.

**Pull request merged**